lib/poparser.js

```diff
@@ -3,53 +3,95 @@ const sharedFuncs = require('./shared');
 };
 
 /**
- * Parses a PO stream, emits translation table in object mode
-
- * @param defaultCharset Default charset to use
- * @param options Stream options
- * @param validation Enables validation of parsed content, if set to true
+ * Normalizes arguments passed into `stream` function
+ *
+ * @typedef {{ defaultCharset?: string, validation?: boolean }} Options
+ * @typedef {{ transformOptions: import('readable-stream').TransformOptions, defaultCharset?: string, validation?: boolean }} ReturnType
  * @type {{
- *    () => Stream;
- *    (defaultCharset: string) => Stream;
- *    (defaultCharset: string, options: import('readable-stream').TransformOptions) => Stream;
- *    (defaultCharset: string, options: import('readable-stream').TransformOptions, validation: boolean) => Stream;
- *    (defaultCharset: string, validation: boolean) => Stream;
- *    (options: import('readable-stream').TransformOptions) => Stream;
- *    (options: import('readable-stream').TransformOptions, validation: boolean) => Stream;
- *    (validation: boolean) => Stream;
+ *    () => ReturnType;
+ *    (defaultCharset: string) => ReturnType;
+ *    (defaultCharset: string, transformOptions: import('readable-stream').TransformOptions) => ReturnType;
+ *    (transformOptions: import('readable-stream').TransformOptions) => ReturnType;
+ *    (options: Options) => ReturnType;
+ *    (options: Options, transformOptions: import('readable-stream').TransformOptions) => ReturnType;
  * }}
  */
-module.exports.stream = function (defaultCharset, options, validation) {
-  const args = sharedFuncs.unshiftOptionalArgs(
-    ['string', 'object', 'boolean'],
-    [defaultCharset, options, validation]
-  );
+function normalizeStreamArgs (arg0 = {}, arg1 = {}) {
+  return typeof arg0 === 'string'
+    ? { defaultCharset: arg0, transformOptions: arg1 }
+    : 'defaultCharset' in arg0 || 'validation' in arg0
+      ? { defaultCharset: arg0.defaultCharset, validation: arg0.validation, transformOptions: arg1 }
+      : { transformOptions: arg0 };
+}
 
-  return new PoParserTransform(...args);
+/**
+ * Parses a PO stream, emits translation table in object mode
+ *
+ * @typedef {{ defaultCharset: strubg, validation: boolean }} Options
+ * @param defaultCharset Default charset to use
+ * @param transformOptions Stream options
+ * @param options Optional options with defaultCharset and validation
+ * @type {{
+*    () => Stream;
+*    (defaultCharset: string) => Stream;
+*    (defaultCharset: string, transformOptions: import('readable-stream').TransformOptions) => Stream;
+*    (transformOptions: import('readable-stream').TransformOptions) => Stream;
+*    (options: Options) => Stream;
+*    (options: Options, transformOptions: import('readable-stream').TransformOptions) => Stream;
+* }}
+*/
+function stream (arg0, arg1) {
+  const { defaultCharset, validation, transformOptions } = normalizeStreamArgs(arg0, arg1);
+
+  return new PoParserTransform(defaultCharset, transformOptions, validation);
```

Comment:
Let's use the same pattern here:
```js
PoParserTransform(options, transformOptions);
```
This means `function stream` above does not need anything, just pass on the arguments.

`PoParserTransform` itself would then need to become something like:

```js
function PoParserTransform (options, transformOptions) {
  if (typeof options === 'string') {
    options = { defaultCharset: options };
  }

  const { defaultCharset, validation } = options;
```

I hope this works. Feel free to say otherwise.
The idea is this would eliminated the need for `normalizeStreamArgs` as there is only 1 special use case (the old, `defaultCharset`) and that should not obfuscate the recommended new usage. Although anyone who cares can clearly see it this way (and of course, please keep the remaining type definitions in jsdoc comments - thanks for those!).

Reply:
@smhg I hope you don't mind if share my motivation for solving it the way I did it.

It is partially described in https://github.com/smhg/gettext-parser/pull/76#discussion_r1136781289 and, to add to that, I believe this is a good occasion to apply single responsibility principle. I think that parsing incoming arguments should be decoupled from creating an instance of a class. That is why I decided to introduce arguments normalization functions and keep the constructor of `PoParserTransform` a bit thinner. Because those normalization functions are pure I, subjectively, find it to be closer to functional programming style (which, according to our discussion in the relevant [issue](https://github.com/smhg/gettext-parser/issues/75#issuecomment-1344327213), I believe is desirable). Additionally. as a consequence of such decoupling we are able to easily provide fine grained unit test coverage. If we decide to couple those two things back together, I believe testing that would become a bit less convenient.

Given the above, what do you think about it?

Reply:
https://github.com/smhg/gettext-parser/pull/76#discussion_r1144566489